proposals/nnnn-trailing-comma-lists.md

```diff
@@ -0,0 +1,294 @@
+for (a, b,) in zip(s1, s2) { }
+```
+
+**Trailing comma will be allowed in single-element lists but not in zero-element lists.**
+
+```swift
+(1,) // OK
+(,) // ❌ expected value in tuple
+```
+
+- Initializers, functions and associated values arguments and parameters
+
+```swift
+
+m[x, y,]
+
+func foo(a: Int, b: Int,) { }
+
+struct S {
+    init(a: Int, b: Int,) { }
+}
+
+enum E {
+    case foo(a: Int, b: Int,)
+}
+```
+- Subscripts
+
+```swift
+let value = m[x, y,]
+```
+
+- `KeyPath` subscripts 
+
+```swift
+let keyPath = \Foo.bar[x,y,]
+          
+f(\.[x,y,])
+```
+
+- Attributes arguments
+
+```swift
+@Foo(1, 2, 3,) struct S { }
+f(_: @foo(1, 2,) Int)
+```
+
+- Macro expansion arguments
+
+```swift
+#foo(1, 2,)
+
+struct S {
+    #foo(1, 2,)
+}
+```
+
+- `if`, `guard` and `while` conditions
+
+```swift
+if a, b, { }
+while a, b, { }
+guard a, b, else { }
+```
+
+- `switch` case labels
+
+```swift
+switch number {
+    case 1, 2,:
+        ...
+    default:
+        ..
+}
+```
+
+- Closure capture list
+
+```swift
+{ [a, b,] in }
+```
+
+- inheritance list
+
+```swift
+struct S: P1, P2, P3, { }
+```
+
+**Inheritance lists won't support trailing comma when there isn't a clear terminator, such as `associatedtype` in a protocol declaration:**
+```swift
+protocol Foo {
+  associatedtype T: P1, P2, // ❌ Expected type
+}
+```
+
+- Generic parameters
+
+```swift
+struct S<T1, T2, T3,> { }
+```
+
+- Generic `where` clause list
+
+```swift
+struct S<T1, T2, T3> where T1: P1, T2: P2, { }
+```
+
+**Generic `where` clause lists won't support trailing comma when there isn't a clear terminator, such as functions in protocol declaration:**
+```swift
+protocol Foo {
+  func f<T1, T2>(a: T1, b: T2) where T1: P1, T2: P2, // ❌ Expected type
+}
+```
+
+- Availability spec list
+
+```swift
+if #unavailable(iOS 15, watchOS 9,) { }
+```
+
+- String interpolation
+
+```swift
+let s = "\(1, 2,)"
+```
+
+## Source compatibility
+
+Although this change won't impact existing valid code it will change how some invalid codes are parsed. Consider the following:
+
+```swift
+if
+  condition1,
+  condition2,
+{ // ❌ Function produces expected type 'Bool'; did you mean to call it with '()'?
+  return true
+} 
+
+{ print("something") }
+```
+
+Currently the parser uses the last comma to determine that whatever follows is the last condition, so `{ return true }` is a condition and `{ print("something") }` is the if body.
+To allow trailing comma the proposed solution is to change the parser to terminate de condition list before the first block that is a valid if body, so `{ return true }` will the parsed as the if body.
+
+```swift
+if
+  condition1,
+  condition2,
+{
+  return true
+} 
+
+{ print("something") } // ❌ Closure expression is unused
+```
+
+## Alternatives considered
+
+### Eliding commas
+
+A different approach to address the exact same motivation is to allow the comma between two expressions to be elided when they are separated by a newline.
```

Comment:
```suggestion
A different approach to address similar motivations is to allow the comma between two expressions to be elided when they are separated by a newline.
```

They don't have to be the "exact same" motivation, or at least it's not necessary for you to argue that they're the exact same motivations here.